crates/bevy_ecs/src/component.rs

```diff
@@ -1135,73 +1138,536 @@ impl ComponentCloneBehavior {
     }
 }
 
-/// Stores metadata associated with each kind of [`Component`] in a given [`World`].
+/// A queued component registration.
+struct QueuedRegistration {
+    registrator: Box<dyn FnOnce(&mut ComponentsRegistrator, ComponentId)>,
```

Comment:
I think it might be possible to avoid allocating a `Box` here.  

The component and resource versions could pass plain `fn` pointers, since they don't need to capture anything.  (The resource one seems to capture `type_id::of::<T>()`, but you could evaluate that again in the function.)  

The dynamic component one has a `ComponentDescriptor` value, but always runs the same code.  So I think that could be a `Vec<(ComponentId, ComponentDescriptor)>` and not even have to do virtual dispatch.

Reply:
Yeah, we can wait on this.

Reply:
> If you feel strongly about it

Nope, this is just Rust tempting me to micro-optimize unimportant things :).